### ISSUE TITLE
Runtime Planning

### ISSUE THREAD

**Issue by reem** [reem/transfer#1]
This planning issue covers the transfer runtime, which is defined in the `transfer::rt` module.

The transfer runtime is a fairly thin wrapper over the underlying mio API. It wraps an event loop, an allocator, and an "executor" which is most often some form of thread pool. Multiple runtimes can use the same allocator and executor, and in the future the reverse will also be possible. These metadata are wrapped into a single type, `rt::Metadata`, in the implementation.

The runtime currently supports registering next-tick handlers, timeouts, and tcp listeners. In the future, there will be support for general io listeners, so that servers using transfer can use the same event loop, allocator, and executor to handle asynchronous events.

Currently, all communication with the runtime is managed via the `rt::Handle` type, which communicates with the runtime exclusively through the notify queue provided by mio. `rt::Handle` also holds a `Future` which will be completed when the event loop is shut down; this is used for "blocking" on the runtime.

Once started, the runtime will use the event loop to wait for new io events, timeouts, or next-tick handlers. When a new connection is received on a registered listener, an eventual stream is set up to parse and handle requests. The actual parsing and executing occurs in one of the executors "threads."

Reading and parsing is facilitated by `iobuf::AppendBuf` and `iobuf::AROIobuf`. New data is read into an `AppendBuf` and `AROIobuf` snapshots are sent to the stream for this connection, which will attempt to parse them as http requests using `httparse`, and are then handled in the same executor "thread" if they parse correctly.

**Comment by reem** [comment]
Did some thinking and evaluation of the existing space and would like to catalogue some new plans:

Each rt instance will use one io thread, which manages communication with user code (what runs on the executor) through the notify queue, read events, and write events.

User code sends events to the io thread through the notify queue, and the io thread sends events to to user code through read and write events, which will either trigger a new request or send read/write evidence through.

In order to tackle #2 and #3 efficiently, I am considering an abstraction similar to the one prototyped in [`chunks`](https://github.com/reem/rust-chunks), based on a `Stream<ReadEvidence>` and `Stream<WriteEvidence>`. The non-buffered nature of streams is ideal in this case, since you can't queue evidence.

In order to avoid the overhead of sending snapshots and the excessively awkward dance of converting between slices and AROIobufs, we do the parsing of requests _on the io thread_. Since `httparse` is blazingly fast, this can add only a few hundred nanoseconds of latency, but if it turns out to be a performance problem it is possible to re-adjust the scheme to send snapshots, as is done today.

To summarize:

The io thread is responsible for:
- Reading the request head
- Parsing requests (just the head!)
- Managing registrations
- Writing the response head
- Triggering next-tick and timeout handlers

User code is responsible for:
- Reading the request body (through a chunks-like stream-of-evidence abstraction)
- Writing the response body (through a chunks-like stream-of-evidence abstraction)
- Handling requests (can block)
- Running timeout and next-tick handlers (can block)
- Hosting the io thread (the io thread is also scheduled on the passed-in Executor)

**Comment by reem** [comment]
Also, the runtime should support both edge and level polling as a configurable option - this means the code for reading the request head and writing the response head must work under both edge and level conditions. Outside of the io thread, user code is then responsible for not stalling out the evidence stream by reading/writing only some of the data if it is using edge polling.
